helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/model/ClusterModelProvider.java

```diff
@@ -391,6 +418,46 @@ private static Set<AssignableReplica> findToBeAssignedReplicasByClusterChanges(
     return toBeAssignedReplicas;
   }
 
+  /**
+   * Find replicas that were assigned to non-active nodes in the current assignment.
+   *
+   * @param replicaMap             A map contains all the replicas grouped by resource name.
+   * @param activeInstances        All the instances that are live and enabled according to the delay rebalance configuration.
+   * @param currentAssignment      The current assignment that was generated in the previous rebalance.
+   * @param allocatedReplicas      A map of <Instance -> replicas> to return the allocated replicas grouped by the target instance name.
+   * @return The replicas that need to be reassigned.
+   */
+  private static Set<AssignableReplica> findToBeAssignedReplicasOnDownInstances(
+      Map<String, Set<AssignableReplica>> replicaMap, Set<String> activeInstances,
+      Map<String, ResourceAssignment> currentAssignment,
+      Map<String, Set<AssignableReplica>> allocatedReplicas) {
+    // For any replica that are assigned to non-active instances (down instances), add them.
+    Set<AssignableReplica> toBeAssignedReplicas = new HashSet<>();
+    for (String resourceName : replicaMap.keySet()) {
+      Map<String, Map<String, Set<String>>> stateInstanceMap = getStateInstanceMap(currentAssignment.get(resourceName));
+
+      for (AssignableReplica replica : replicaMap.get(resourceName)) {
+        String partitionName = replica.getPartitionName();
+        String replicaState = replica.getReplicaState();
+        Set<String> currentAllocations =
+            stateInstanceMap.getOrDefault(partitionName, Collections.emptyMap())
+                .getOrDefault(replicaState, Collections.emptySet());
+        if (!currentAllocations.isEmpty()) {
+          String allocatedInstance = currentAllocations.iterator().next();
+          if (activeInstances.contains(allocatedInstance)) {
+            allocatedReplicas.computeIfAbsent(allocatedInstance, key -> new HashSet<>()).add(replica);
```

Comment:
for my own understanding, do we only check the first allocatedInstance? Any chance there is more than one element in the set?

Reply:
Since the loop is based on per replica, we're checking instances one at a time; there could be 2 instances in the currentAllocations set, but this set will also be visited 2 times if there're 2 FOLLOWER replica.

Reply:
then I'm confused, the code is using `currentAllocations.iterator().next();` each time, from your description it should be reusing the same iterator.

Reply:
That's because `currentAllocations` is losing one element each time a corresponding replica is there, if you look a few lines down. So `currentAllocations.iterator().next()` is just the first element (whatever that means in a set) of this set.

Reply:
I see the logic here, thanks for the elaboration. 
One thing to improve is not to remove and add element one-by-one, but in one pass for each resource. Not sure if that's applicable at all, so just nice to have. Feel free to close this.